File: cpp/perceptron/and_gate.cpp
#include <iostream>
#include <vector>
#include "Perceptron.h"
#include "activation.h"

int main() {
    std::cout << "=== AND Gate Implementation with Perceptron ===" << std::endl << std::endl;
    
    // Create training data for AND gate
    std::vector<std::vector<double>> X = {
        {0.0, 0.0},
        {0.0, 1.0},
        {1.0, 0.0},
        {1.0, 1.0}
    };
    
    std::vector<int> y = {0, 0, 0, 1};  // AND gate truth table
    
    // Initialize perceptron with 2 inputs
    Perceptron perceptron(2, 0.1, 100, 42);
    
    // Train the perceptron
    std::cout << "Training perceptron on AND gate..." << std::endl;
    perceptron.fit(X, y);
    
    // Display learned weights and bias
    std::cout << "Training completed in " << perceptron.get_errors().size() << " epochs." << std::endl;
    std::cout << "Learned weights: [" << perceptron.get_weights()[0] << ", " 
              << perceptron.get_weights()[1] << "]" << std::endl;
    std::cout << "Learned bias: " << perceptron.get_bias() << std::endl << std::endl;
    
    // Test the perceptron
    std::cout << "Testing AND gate:" << std::endl;
    std::cout << "----------------" << std::endl;
    
    std::vector<int> predictions = perceptron.predict(X);
    
    for (size_t i = 0; i < X.size(); ++i) {
        std::cout << "Input: [" << X[i][0] << ", " << X[i][1] << "] -> Output: " << predictions[i] 
                  << " (Expected: " << y[i] << ")" << std::endl;
    }
    
    return 0;
}

File: rust/perceptron/src/bin/and_gate.rs
use perceptron::perceptron::Perceptron;

fn main() {
    println!("=== AND Gate Implementation with Perceptron ===\n");
    
    // Create training data for AND gate
    let x = vec![
        vec![0.0, 0.0],
        vec![0.0, 1.0],
        vec![1.0, 0.0],
        vec![1.0, 1.0],
    ];
    
    let y = vec![0, 0, 0, 1];  // AND gate truth table
    
    // Initialize perceptron with 2 inputs
    let mut perceptron = Perceptron::new(2, 0.1, 100, 42);
    
    // Train the perceptron
    println!("Training perceptron on AND gate...");
    perceptron.fit(&x, &y);
    
    // Display learned weights and bias
    println!("Training completed in {} epochs.", perceptron.get_errors().len());
    print!("Learned weights: [");
    for (i, &weight) in perceptron.get_weights().iter().enumerate() {
        if i > 0 {
            print!(", ");
        }
        print!("{}", weight);
    }
    println!("]");
    println!("Learned bias: {}\n", perceptron.get_bias());
    
    // Test the perceptron
    println!("Testing AND gate:");
    println!("----------------");
    
    let predictions = perceptron.predict(&x);
    
    for (i, sample) in x.iter().enumerate() {
        println!(
            "Input: [{}, {}] -> Output: {} (Expected: {})",
            sample[0], sample[1], predictions[i], y[i]
        );
    }
}
